NDTensors/src/linearalgebra/linearalgebra.jl

```diff
@@ -388,23 +367,122 @@ function LinearAlgebra.eigen(
   return D, V, spec
 end
 
-function LinearAlgebra.qr(T::DenseTensor{ElT,2,IndsT}; kwargs...) where {ElT,IndsT}
-  positive = get(kwargs, :positive, false)
-  # TODO: just call qr on T directly (make sure
-  # that is fast)
-  if positive
-    QM, RM = qr_positive(matrix(T))
-  else
-    QM, RM = qr(matrix(T))
-  end
+function qr(T::DenseTensor{<:Any,2,IndsT}; positive=false, kwargs...) where {IndsT}
+  qxf = positive ? qr_positive : qr
+  return qx(qxf, T; kwargs...)
+end
+function ql(T::DenseTensor{<:Any,2,IndsT}; positive=false, kwargs...) where {IndsT}
+  qxf = positive ? ql_positive : ql
+  return qx(qxf, T; kwargs...)
+end
+
+#
+#  Generic function for qr and ql decomposition of dense matrix.
+#  The X tensor = R or L.
+#
+function qx(qx::Function, T::DenseTensor{<:Any,2,IndsT}; kwargs...) where {IndsT}
+  QM, XM = qx(matrix(T))
   # Make the new indices to go onto Q and R
   q, r = inds(T)
   q = dim(q) < dim(r) ? sim(q) : sim(r)
   Qinds = IndsT((ind(T, 1), q))
-  Rinds = IndsT((q, ind(T, 2)))
-  Q = tensor(Dense(vec(Matrix(QM))), Qinds)
-  R = tensor(Dense(vec(RM)), Rinds)
-  return Q, R
+  Xinds = IndsT((q, ind(T, 2)))
+  Q = tensor(Dense(vec(Matrix(QM))), Qinds) #Q was strided
+  X = tensor(Dense(vec(XM)), Xinds)
+  return Q, X
+end
+
+#
+# Just flip signs between Q and R to get all the diagonals of R >=0.
+# For rectangular M the indexing for "diagonal" is non-trivial.
+#
+"""
+    qr_positive(M::AbstractMatrix)
+
+Compute the QR decomposition of a matrix M
+such that the diagonal elements of R are
+non-negative. Such a QR decomposition of a
+matrix is unique. Returns a tuple (Q,R).
+"""
+function qr_positive(M::AbstractMatrix)
+  sparseQ, R = qr(M)
+  Q = convert(Matrix, sparseQ)
+  nc = size(Q, 2)
+  for c in 1:nc
+    sign_Rc = sign(R[c, c])
+    if !isone(sign_Rc)
+      R[c, c:end] *= conj(sign_Rc) #only fip non-zero portion of the row.
+      Q[:, c] *= sign_Rc
+    end
+  end
+  return (Q, R)
+end
+
+"""
+    ql_positive(M::AbstractMatrix)
+
+Compute the QL decomposition of a matrix M
+such that the diagonal elements of L are
+non-negative. Such a QL decomposition of a
+matrix is unique. Returns a tuple (Q,L).
+"""
+function ql_positive(M::AbstractMatrix)
+  sparseQ, L = ql(M)
+  Q = convert(Matrix, sparseQ)
+  nr, nc = size(L)
+  dc = nc > nr ? nc - nr : 0 #diag is shifted over by dc if nc>nr
+  for c in 1:(nc - dc)
+    sign_Lc = sign(L[c, c + dc])
+    if c <= nr && !isone(sign_Lc)
+      L[c, 1:(c + dc)] *= sign_Lc #only fip non-zero portion of the column.
+      Q[:, c] *= conj(sign_Lc)
+    end
+  end
+  return (Q, L)
+end
+
+#
+#  Lapack replaces A with Q & R carefully packed together.  So here we just copy a
+#  before letting lapack overwirte it. 
+#
+function ql(A::AbstractMatrix{T}; kwargs...) where {T}
```

Comment:
Can drop `T` and `where {T}` in the function signature and use `eltype(A)` instead of `T` inside the function.

Reply:
Done